shared/actions/signup.js

```diff
@@ -1,366 +1,198 @@
+  RPCTypes.deviceCheckDeviceNameFormatRpcPromise({name: state.signup.devicename}, Constants.waitingKey)
+    .then(() => SignupGen.createCheckedDevicename({devicename: action.payload.devicename}))
+    .catch(error =>
+      SignupGen.createCheckedDevicenameError({
+        devicename: action.payload.devicename,
+        error: `Device name is invalid: ${error.desc}.`,
+      })
+    )
 
-              const signupPromise = dispatch(signup(skipMail || false, onDisplayPaperKey))
-              if (signupPromise) {
-                signupPromise.then(resolve, reject)
-              } else {
-                throw new Error('did not get promise from signup')
-              }
-            }
-          })
-          .catch(err => {
-            logger.warn('device name is invalid: ', err)
-            dispatch(
-              SignupGen.createSubmitDeviceNameError({
-                deviceNameError: `Device name is invalid: ${err.desc}.`,
-              })
-            )
-            resolve()
-          })
-      }
-    })
-    return p
+// Actually sign up ///////////////////////////////////////////////////////////
+const reallySignupOnNoErrors = (_, state: TypedState) => {
+  if (!noErrors(state)) {
+    logger.warn('Still has errors, bailing on really signing up')
+    return
   }
-}
 
-let paperKeyResponse = null
-function sawPaperKey() {
-  return () => {
-    if (paperKeyResponse) {
-      paperKeyResponse.result()
-      paperKeyResponse = null
-    }
+  const {email, username, inviteCode, passphrase, devicename} = state.signup
+
+  if (!email || !username || !inviteCode || !passphrase || !passphrase.stringValue() || !devicename) {
+    logger.warn(
+      'Missing data during signup phase',
+      email,
+      username,
+      inviteCode,
+      devicename,
+      !!passphrase,
+      passphrase && !!passphrase.stringValue()
+    )
+    throw new Error('Missing data for signup')
   }
-}
 
-function signup(skipMail: boolean, onDisplayPaperKey?: () => void) {
-  return (dispatch, getState) => {
-    const p = new Promise((resolve, reject) => {
-      const {email, username, inviteCode, passphrase, deviceName} = getState().signup
-      paperKeyResponse = null
-      const deviceType = isMobile ? RPCTypes.commonDeviceType.mobile : RPCTypes.commonDeviceType.desktop
-
-      if (email && username && inviteCode && passphrase && deviceName) {
-        RPCTypes.signupSignupRpcPromise({
-          incomingCallMap: {
-            'keybase.1.gpgUi.wantToAddGPGKey': (params, response) => {
-              // Do not add a gpg key for now
-              response.result(false)
-            },
-            'keybase.1.loginUi.displayPrimaryPaperKey': ({sessionID, phrase}, response) => {
-              paperKeyResponse = response
-              dispatch(SignupGen.createShowPaperKey({paperkey: new HiddenString(phrase)}))
-              onDisplayPaperKey && onDisplayPaperKey()
-              dispatch(nextPhase())
-            },
+  return Saga.call(function*() {
```

Comment:
We usually encode this pattern w/ `safeTakeEveryPure(..., actionCreatorsWithError)`, could you do that here?